### PR TITLE
fix: Added Batch operations.

### PR DESCRIPTION
**Pull request by DivyPatel9881**
fix: #21

### CODE REVIEW

File: src/adapter.ts [hunk 1]
```diff
@@ -143,6 +143,19 @@ export default class TypeORMAdapter implements Adapter {
...
+     * addPolicies adds policy rules to the storage.
+     */
+    public async addPolicies(sec: string, ptype: string, rules: string[][]) {
+        const lines: GenericCasbinRule[] = [];
+        for (const rule of rules) {
+            const line = this.savePolicyLine(ptype, rule);
+            lines.push(line);
+        }
+
+        await getRepository(this.getCasbinRuleConstructor(), this.option.name).save(lines);
```
Comment (nodece) [inline src/adapter.ts:156]: It is important to note that most databases have the maximum batch size, If the batch size is too large will fail.

I'm not sure the TypeORM how to deal with.

File: src/adapter.ts [hunk 2]
```diff
@@ -151,6 +164,17 @@ export default class TypeORMAdapter implements Adapter {
         await getRepository(this.getCasbinRuleConstructor(), this.option.name).delete(line);
     }
 
+    /**
+     * removePolicies removes policy rules from the storage.
+     */
+    public async removePolicies(sec: string, ptype: string, rules: string[][]) {
+        const repository = getRepository(this.getCasbinRuleConstructor(), this.option.name);
```
Comment (nodece) [inline src/adapter.ts:171]: +1

**Comment by nodece** [comment]
Could you add transaction in `addPolicies`, `removePolicies` and `savePolicy`? 

When happens error, rollback and throw error.

**Comment by nodece** [comment]
Before PR,  make commits is clean via squash and rebase.

**Comment by hsluoyz** [comment]
@nodece

File: src/adapter.ts [hunk 3]
```diff
@@ -143,6 +156,31 @@ export default class TypeORMAdapter implements Adapter {
...
+        const queryRunner = this.typeorm.createQueryRunner();
+
+        await queryRunner.connect();
+        await queryRunner.startTransaction();
+
+        try {
+            await queryRunner.manager.save(lines);
+            await queryRunner.commitTransaction();
+        } catch (err) {
+            await queryRunner.rollbackTransaction();
```
Comment (nodece) [inline src/adapter.ts:178]: When an error occurs, we need to throw it

File: src/adapter.ts [hunk 4]
```diff
@@ -151,6 +189,27 @@ export default class TypeORMAdapter implements Adapter {
...
+        await queryRunner.connect();
+        await queryRunner.startTransaction();
+
+        try {
+            for (const rule of rules) {
+                const line = this.savePolicyLine(ptype, rule);
+                await getRepository(this.getCasbinRuleConstructor(), this.option.name).delete(line);
+            }
+        } catch (err) {
+            await queryRunner.rollbackTransaction();
```
Comment (nodece) [inline src/adapter.ts:207]: throw the err.

File: src/adapter.ts [hunk 5]
```diff
@@ -151,6 +189,27 @@ export default class TypeORMAdapter implements Adapter {
...
+    public async removePolicies(sec: string, ptype: string, rules: string[][]) {
+        const queryRunner = this.typeorm.createQueryRunner();
+
+        await queryRunner.connect();
+        await queryRunner.startTransaction();
+
+        try {
+            for (const rule of rules) {
+                const line = this.savePolicyLine(ptype, rule);
+                await getRepository(this.getCasbinRuleConstructor(), this.option.name).delete(line);
```
Comment (nodece) [inline src/adapter.ts:204]: It seems should use the queryRunner.

File: src/adapter.ts [hunk 6]
```diff
@@ -151,6 +191,29 @@ export default class TypeORMAdapter implements Adapter {
...
+        const type = TypeORMAdapter.getCasbinRuleType(this.option.type);
+
+        await queryRunner.connect();
+        await queryRunner.startTransaction();
+
+        try {
+            for (const rule of rules) {
+                const line = this.savePolicyLine(ptype, rule);
+                await queryRunner.manager.delete(type, line);
+            }
```
Comment (nodece) [inline src/adapter.ts:208]: miss `await queryRunner.commitTransaction()`

**Comment by DivyPatel9881** [comment]
Thanks for pointing out, @nodece, so silly of me.

**Comment by nodece** [comment]
lgtm

**Pull request merged**